### PR TITLE
60 Memory error when creating fine anisotropic grid

### PR DESCRIPTION
**Pull request by CodingAllan**
## Merge Request - GuideLine Checklist 

Implemented small fix where iterator could go out of bounds in creating anisotropic grid. This was the case, when nr_exp=fac_ani+1 and fac_ani>3 with alpha_coeff either being 0 or 2.

**Guideline** to check code before resolve WIP and approval, respectively.
As many checkboxes as possible should be ticked.

### Checks by code author:
Always to be checked:
* [x] There is at least one issue associated with the pull request.
* [x] New code adheres with the [coding guidelines](https://github.com/mknaranja/GMGPolar/wiki)
* [x] No large data files have been added to the repository. Maximum size for files should be of the order of KB not MB. In particular avoid adding of pdf, word, or other files that cannot be change-tracked correctly by git.

If functions were changed or functionality was added:
* [ ] Tests for new functionality has been added
* [ ] A local test was succesful

If new functionality was added:
* [ ] There is appropriate **documentation** of your work. (use doxygen style comments)

If new third party software is used:
* [x] Did you pay attention to its license? Please remember to add it to the wiki after successful merging.

If new mathematical methods or epidemiological terms are used:
* [x] Are new methods referenced? Did you provide further documentation?

[ ] The following questions are addressed in the documentation (if need be): 
* Developers (what did you do?, how can it be maintained?)
* For users (how to use your work?)
* For admins (how to install and configure your work?)

* For documentation: Please write or update the Readme in the current working directory!

### Checks by code reviewer(s):
* [x] Is the code clean of development artifacts e.g., unnecessary comments, prints, ...
* [x] The ticket goals for each associated issue are reached or problems are clearly addressed (i.e., a new issue was introduced).
* [x] There are appropriate **unit tests** and they pass.
* [x] The git history is clean and linearized for the merge request. All reviewers should squash commits and write a simple and meaningful commit message.
* [x] Coverage report for new code is acceptable. 
* [x] No large data files have been added to the repository. Maximum size for files should be of the order of KB not MB. In particular avoid adding of pdf, word, or other files that cannot be change-tracked correctly by git.

### CODE REVIEW

**Comment by mknaranja** [comment]
@CodingAllan Please try to give it a simple (as with the issue) and provide details in the description ;-)

File: src/create_grid_polar.cpp [hunk 1]
```diff
@@ -142,28 +142,36 @@ void level::build_r()
...
+        }
+        else if (gyro::icntl[Param::alpha_coeff] == POISSON) {
             r_jump = 0.5; // There is no jump for Poisson so this is an arbitrary choice
-        } else {
+        }
+        else {
             throw std::runtime_error("Unknown alpha coeff");
         }
-        se = floor(nr * r_jump) - n_elems_refined / 2;
+        if (floor(nr * r_jump) > nr - (n_elems_refined / 2)) {
```
Comment (mknaranja) [inline src/create_grid_polar.cpp:170]:
```suggestion
        // correct anisotropy setting if value was too large for chosen discretization in r
        if (floor(nr * r_jump) > nr - (n_elems_refined / 2)) {
            std::cout << "Combination for (nr, anisotropy) was not suitable, anisotropy value has been reduced to " << new_aniso  << "." << std::endl;
```

**Review by mknaranja — CHANGES_REQUESTED** [review]
Thanks for the fix. Please check and apply my suggestions if you agree.

File: src/create_grid_polar.cpp [hunk 2]
```diff
@@ -142,28 +142,36 @@ void level::build_r()
...
+            int new_aniso   = log2(nr - floor(nr * r_jump)) + 1;
```
Comment (mknaranja) [inline src/create_grid_polar.cpp:171]: I think there is another error introduced since the old aniso is still used in e.g. line 188 etc

**Comment by mknaranja** [comment]
Hi @CodingAllan, I have now seen that there was a code shortly after my last review. Please request the review once changes are made, verify that all comments are addressed and answer on the opened conversations. Actually it looks like if two comments have been ignored?

**Comment by CodingAllan** [comment]
@mknaranja , yes I just remembered that this was not merged yet and hence requested a review.

I commented on the suggestions you made and just wanted to ask if the way of correcting is acceptable.
 Im not sure if I am missing something.. but I did not ignore anything ?

**Comment by CodingAllan** [comment]
> @mknaranja , yes I just remembered that this was not merged yet and hence requested a review.
> 
> I commented on the suggestions you made and just wanted to ask if the way of correcting is acceptable. Im not sure if I am missing something.. but I did not ignore anything ?

I am sorry, I accidentally marked a conversation as resolved.

**Review by mknaranja — APPROVED** [review]

**Pull request merged**